napari/layers/graph/graph.py

```diff
@@ -396,13 +397,45 @@ def add(
         """Adds nodes at coordinates.
         Parameters
         ----------
-        coords : sequence of indices to add point at
+        coords : sequence of coordinates for each new node.
         indices : optional indices of the newly inserted nodes.
         """
+        if indices is None:
+            count_adding = len(np.atleast_2d(coords))
+            indices = self.data.get_next_valid_indices(count_adding)
+        indices = np.atleast_1d(indices)
+        if indices.ndim > 1:
+            raise ValueError(
+                trans._(
+                    "Indices for removal must be 1-dim. Found {ndim}",
+                    ndim=indices.ndim,
+                )
+            )
+
+        self.events.data(
+            value=self.data,
+            action=ActionType.ADDING,
+            data_indices=tuple(indices),
+            vertex_indices=((),),
+        )
+
         prev_size = self.data.n_allocated_nodes
-        self.data.add_nodes(indices=indices, coords=coords)
+        added_indices = self.data.add_nodes(indices=indices, coords=coords)
         self._data_changed(prev_size)
 
+        self.events.data(
+            value=self.data,
+            action=ActionType.ADDED,
+            data_indices=tuple(
+                added_indices,
+            ),
+            vertex_indices=((),),
+        )
+        buffer_indices = {
+            self.data._world2buffer[idx] for idx in added_indices
```

Comment:
you can replace this loop with `self.data._map_world2buffer(added_indices)` for a numba accelerated version.